src/d3d12/d3d12_descriptors_allocations.cpp

```diff
@@ -0,0 +1,391 @@
+#include "d3d12_descriptors_allocations.hpp"
+#include "d3d12_renderer.hpp"
+#include "d3d12_functions.hpp"
+#include "../util/log.hpp"
+
+///////////////////////////
+// DESCRIPTOR ALLOCATION //
+///////////////////////////
+
+namespace wr
+{
+
+	DescriptorAllocation::DescriptorAllocation()
+		: m_descriptor{ 0 }
+		, m_num_handles(0)
+		, m_descriptor_size(0)
+		, m_page(nullptr)
+	{}
+
+	DescriptorAllocation::DescriptorAllocation(d3d12::DescHeapCPUHandle descriptor,
+		uint32_t num_handles,
+		uint32_t descriptor_size,
+		std::shared_ptr<DescriptorAllocatorPage> page)
+		: m_descriptor(descriptor)
+		, m_num_handles(num_handles)
+		, m_descriptor_size(descriptor_size)
+		, m_page(page)
+	{}
+
+
+	DescriptorAllocation::~DescriptorAllocation()
+	{
+		Free();
+	}
+
+	DescriptorAllocation::DescriptorAllocation(DescriptorAllocation&& allocation)
+		: m_descriptor(allocation.m_descriptor)
+		, m_num_handles(allocation.m_num_handles)
+		, m_descriptor_size(allocation.m_descriptor_size)
+		, m_page(std::move(allocation.m_page))
+	{
+		allocation.m_descriptor.m_native.ptr = 0;
+		allocation.m_num_handles = 0;
+		allocation.m_descriptor_size = 0;
+	}
+
+	DescriptorAllocation& DescriptorAllocation::operator=(DescriptorAllocation&& other)
+	{
+		// Free this descriptor if it points to anything.
+		Free();
+
+		m_descriptor = other.m_descriptor;
+		m_num_handles = other.m_num_handles;
+		m_descriptor_size = other.m_descriptor_size;
+		m_page = std::move(other.m_page);
+
+		other.m_descriptor.m_native.ptr = 0;
+		other.m_num_handles = 0;
+		other.m_descriptor_size = 0;
+
+		return *this;
+	}
+
+	void DescriptorAllocation::Free()
+	{
+		if (!IsNull() && m_page)
+		{
+			m_page->Free(std::move(*this));
+			m_descriptor.m_native.ptr = 0;
+			m_num_handles = 0;
+			m_descriptor_size = 0;
+			m_page.reset();
+		}
+	}
+
+	// Check if this a valid descriptor.
+	bool DescriptorAllocation::IsNull() const
+	{
+		return m_descriptor.m_native.ptr == 0;
+	}
+
+	// Get a descriptor at a particular offset in the allocation.
+	d3d12::DescHeapCPUHandle DescriptorAllocation::GetDescriptorHandle(uint32_t offset) const
+	{
+		if (offset > m_num_handles)
+		{
+			LOGC("Specified offset is greater than the number of handles in this Descriptor Allocation");
+		}
+
+		return { m_descriptor.m_native.ptr + (m_descriptor_size * offset) };
+	}
+
+	uint32_t DescriptorAllocation::GetNumHandles() const
+	{
+		return m_num_handles;
+	}
+
+	std::shared_ptr<DescriptorAllocatorPage> DescriptorAllocation::GetDescriptorAllocatorPage() const
+	{
+		return m_page;
+	}
+}
+
+
+///////////////////////////////
```

Comment:
You can use normal doxygen qt style comments as explained in the wiki.

Reply:
👍